### DISCUSSION TITLE
Getting a method from an assembly via Mono.Cecil

### DISCUSSION THREAD

**Discussion by vladislav-miroshnikov** [jbevain/cecil#742] [Q&A] [answered]
Hello everyone! I need a little help getting the method directly from assembly using Mono.Cecil. I need to write a program that would take an assembly (.exe or .dll) as input and replace the operation of addition of decimal numbers with subtraction in the IL code. At the moment I am doing something like this:
var sub_op = typeof (Decimal) .GetMethod ("op_Substraction");
var sub _op _ref = assembly.MainModule.ImportReference (sup_op);
And then I insert the necessary instruction in the right place.
But here's the problem. My program is on .NET Core, and both .NET Core assemblies and .NET Framework can be input.
And if I submit an assembly to the .NET Framework input, and then I receive the required sub_op method, then it is taken through the Reflection API of .NET Core, then when inserted into the supplied assembly, it stops working. This is as it should be, because .NET Core and .NET Framework have different Reflection APIs.
Question: can I get the method I need to subtract decimal numbers not through Reflection, but through Mono.Cecil from the input assembly? Unfortunately, I did not find such a way, so I decided to ask you for help.
And one more small question, do I understand correctly that Mono.Cecil cannot read an .exe file on .Net Core? Dll only?
I would be immensely grateful to you if you would help me.

**Comment by jbevain** [comment, marked as answer]
Hi!

You're completely right, if you're using the `System.Reflection` based overloads of `ImportReference` you're creating references to the currently running .NET framework.

You need to use the overloads of `ImportReference` taking Cecil types, as you can see in those tests:

https://github.com/jbevain/cecil/blob/master/Test/Mono.Cecil.Tests/ImportCecilTests.cs

To answer your last question, Cecil can read any managed exe or dll for .NET Framework or .NET Core, no problem.

**Reply by epamX** [reply]
Hi!
Maybe I'm doing something wrong, but I have the same problem with loading it .exe for _Core 3.1_ and _Net 5.0_ (that files not having, as is known, all the build data). I get the error: 'Format of the executable (. exe) or library (. dll) is invalid.'. 

` using var assembly = AssemblyDefinition.ReadAssembly(filePath, readerParams);`

To be fair, it should be noted that when loading the same file via overrided _AssemblyContex_, it outputs 'Could not load file or assembly'.....Net50.App.exe'. The module was expected to contain an assembly manifest.'. 

In any case, I would like Cecil to handle this internally, because sometimes, when listing files sequentially and examining them, I would like to know the Target version. I apologize for my poor English.

**Reply by jbevain** [reply]
If the file is a native executable and not a ECMA-335 compatible assembly, there's not much Cecil can do.

**Comment by epamX** [comment]
Vladislav, did you manage to process the Net Framework assemblies from the Core application normally? I would be grateful for the sample code, I have not yet managed to properly implement dependencies for the Net Framework from under the Core-injector. Please help me.

**Reply by SimonCropp** [reply]
@epamX if your code accessible, i can take a look
